### PR TITLE
Make Discord id length variable

### PR DESCRIPTION
**Pull request by Markus-Rost**
Fixed assumption of discord ids being always 18 characters long, they are based on a timestamp and older account have shorter ids and future accounts might have even longer ones. There is no good regex for ids besides `[0-9]+` really.

I also improved the matrix mention to discord mention replace to save a regex match. I added `id_as_group` as an argument to `mention_regex()` in case the old functionality without group is needed somewhere. Using `re.finditer()` instead of `re.findall` to get a match object that has the full matched string and not just the group match.

### CODE REVIEW

**Comment by git-bruh** [comment]
Thanks for the PR, i will test and merge it in a few days. I also need to improve the channel lookup:

```
# `except_deleted` for invalid channels.
# TODO can this block for too long ?
- for channel in re.findall(f"<#([0-9]{{{discord.ID_LEN}}})>", content):
+ for channel in re.findall("<#([0-9]+)>", content):
```

Now that the match is broader, someone could spam invalid IDs like `<#1>` and make the bot make ~500  invalid requests (2000 / 4) , so I need to implement a normal cache in a dictionary (from websocket data) that we can look up instead of making a request each time.

**Comment by Markus-Rost** [comment]
Ah yes, I also noticed that an error is thrown on mentioning a channel the bot doesn't have permission to read, which felt weird as the complete channel list (including channel names and description) for a guild is available without needing any permissions. 
So this should be easily solvable by just checking the id with the public channel list of the current guild. For mentions of channels from other guilds it should then just keep the raw mention instead of throwing an error. 

I'll look into changing this to only be a single request to the full channel list instead if multiple requests hitting permission restrictions.

**Comment by Markus-Rost** [comment]
I added a [Get Guild Channels](https://discord.com/developers/docs/resources/guild#get-guild-channels) request to get the guilds full list of channels without running into any permission errors. This will also prevent many mentions from slowing down the bot.

File: appservice/discord.py [hunk 1]
```diff
@@ -107,6 +106,7 @@ def __init__(self, message: dict) -> None:
         self.channel_id = message["channel_id"]
         self.content = message.get("content", "")
         self.id = message["id"]
+        self.guild_id = message.get("guild_id", "")
```
Comment (git-bruh) [inline appservice/discord.py]: If this field is guaranteed to be there, then we should access it directly via `["guild_id"]` so it raises a KeyError if we somehow got an invalid message.

https://discord.com/developers/docs/resources/channel#message-object here the guild_id is marked with a `?` meaning it is optional. But I'm guessing that's only if the message belongs to a DM, any idea about this?

**Review by git-bruh — CHANGES_REQUESTED** [review]

File: appservice/discord.py (hunk 1 above)
Comment (git-bruh) [reply appservice/discord.py]: Sorry for the late response, I had "started" a review, but forgor :skull: that you need to press another button for the review to actually get posted

Comment (Markus-Rost) [reply appservice/discord.py]: Ah yes, guild_id is there for all messages except DMs and Group DMs. Looking at the intents the bot does not receive DMs and as such I'll change it to `["guild_id"]`.

**Review by Markus-Rost — COMMENTED** [review]
I changed your weird hashing of webhook ids which I assume it was to prevent the matrix username from changing for webhook messages with custom name.

The hashing was quite scary as you added the hash value to the webhook id which might be able to result in a valid user id, impersonating other users.

I changed it to appending `-{hashed}` to the matrix userid instead. This also allows for mention conversion to identify webhooks and use the plain text instead of creating a broken Discord mention.

**Comment by git-bruh** [comment]
> The hashing was quite scary as you added the hash value to the webhook id which might be able to result in a valid user id, impersonating other users.

Yeah it was pretty dumb, was just relying on there being a low chance of a collision.  The hash itself is also quite garbage (Just a sum of ascii values, so "abcd" and "dcba" would hash to the same thing) which i had refrained from changing since a ton of new webhooks would be created.

Both of these were implemented by my newbie self for some reason I cannot recall, probably something with ensuring that the ID fits in 18 digits :P I've changed the hash function to a much better one, since webhooks will be recreated now anyways.

**Comment by git-bruh** [comment]
Hmm, seems like webhooks don't have the `guild_id` key set for whatever reason... So sending any message from the bridge breaks it:

```

Traceback (most recent call last):
  File "/home/testuser/git/py/lib/python3.10/site-packages/bottle.py", line 868, in _handle
    return route.call(**args)
  File "/home/testuser/git/py/lib/python3.10/site-packages/bottle.py", line 1748, in wrapper
    rv = callback(*a, **ka)
  File "/home/testuser/git/git/matrix-discord-bridge/tmp/matrix-discord-bridge/appservice/misc.py", line 28, in wrapper
    return fn(self, *args, **kwargs)
  File "/home/testuser/git/git/matrix-discord-bridge/tmp/matrix-discord-bridge/appservice/appservice.py", line 87, in receive_event
    self.handle_event(event)
  File "/home/testuser/git/git/matrix-discord-bridge/tmp/matrix-discord-bridge/appservice/appservice.py", line 66, in handle_event
    func(obj)
  File "/home/testuser/git/git/matrix-discord-bridge/tmp/matrix-discord-bridge/appservice/main.py", line 131, in on_message
    message_id = self.discord.send_webhook(
  File "/home/testuser/git/git/matrix-discord-bridge/tmp/matrix-discord-bridge/appservice/gateway.py", line 232, in send_webhook
    return discord.Message(resp)
  File "/home/testuser/git/git/matrix-discord-bridge/tmp/matrix-discord-bridge/appservice/discord.py", line 109, in __init__
    self.guild_id = message["guild_id"]
KeyError: 'guild_id'
```

I guess this would be a problem with any non-regular discord message aswell, such as ones sent by bots like NQN

**Comment by git-bruh** [comment]
What do you think of fetching the current channel's guild ID and then fetching all channels of that guild ID?

**Comment by Markus-Rost** [comment]
It seems to happen when converting the response from sending with webhook into a Message object.

Actual webhook messages received from ON_MESSAGE likely have a guild_id.

Edit: webhook messages from ON_MESSAGE have guild_id

**Comment by git-bruh** [comment]
Ah yeah, I misread the traceback.

File: appservice/main.py [hunk 2]
```diff
@@ -666,14 +672,16 @@ def process_message(self, message: discord.Message) -> Tuple[str, Dict]:
...
-        for channel in re.findall(f"<#([0-9]{{{discord.ID_LEN}}})>", content):
-            discord_channel = except_deleted(self.get_channel)(channel)
-            name = (
-                discord_channel.name if discord_channel else "deleted-channel"
-            )
-            content = content.replace(f"<#{channel}>", f"#{name}")
+        # Replace channel IDs with names.
+        channels = re.findall("<#([0-9]+)>", content)
+        if channels:
+            discord_channels = self.get_channels(message.guild_id)
```
Comment (git-bruh) [inline appservice/main.py]: Could you add something like this, just in-case?

```
if not message.guild_id:
    self.logger.warning(f"Message '{message.id}' in channel '{message.channel_id}' does not contain a guild_id!")
else:
    # Do mention stuff
```

Comment (Markus-Rost) [reply appservice/main.py]: Added, I also added a comment in Message class.

**Comment by git-bruh** [comment]
Thanks a lot! Sorry for slow responses, shitcord keeps disabling my account asking for phone number so I have to create a bot account once before they do that and use that to ask other users (via [cordless](https://github.com/bios-marcel/cordless)) to send messages so i can test.

**Pull request merged**